File: leetcode/easy/141-linked-list-cycle/141.py
from typing import Optional

class ListNode:
    def __init__(self, val: int = 0, next: 'Optional[ListNode]' = None):
        self.val = val
        self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False

        slow = head
        fast = head

        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next

            if slow is fast:
                return True

        return False

File: leetcode/easy/141-linked-list-cycle/141.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function hasCycle(head: ListNode | null): boolean {
  if (head === null) return false;
  let slow: ListNode | null = head;
  let fast: ListNode | null = head;

  while (fast && fast.next && fast.next.next) {
    slow = slow ? slow.next : null;
    fast = fast.next.next;

    if (slow === fast) return true;
  }

  return false;
}
